monai/transforms/utility/array.py

```diff
@@ -1163,6 +1162,47 @@ def __call__(self, img: NdarrayOrTensor):
         return out
 
 
+class TorchIO(Transform, RandomizableTrait):
```

Comment:
I think we do need this yes, otherwise we're going to explicitly state that only the randomised TorchIO transforms are going to be used here. This probably isn't ideal even though most of the deterministic transforms in TorchIO are represented in native MONAI transforms. Let's have a `TorchIO` and `RandTorchIO` pair of transforms and separate out which are deterministic or not.

Reply:
How would the transform behave with caching if p!=1 in the tio transform [source of tio.Transform](https://github.com/fepegar/torchio/blob/13407ab7f6d562a0953fa6642ce198f297129589/src/torchio/transforms/transform.py#L97C2-L97C22) ?

I get the distinction between probabilistic in terms of application probability (like cropping with p!=1) and randomness inherent in the transform (like random noise) and agree with adding RandTorchIO for the latter case, but am unsure how to deal with the former.

Reply:
We do have a fundamental conflict here in that our transforms differentiate between random ones and deterministic ones, in both type and interface, whereas the interface for TorchIO does not. We may have to rely on the user only using non-random TorchIO transforms with this `TorchIO` class and any randomised ones use `RandTorchIO`. The alternative is that we keep `RandomizableTrait` here and then TorchIO preprocessing transforms essentially can't be used in a deterministic way, ie. can't be cached by `CacheDataset` amongst other things. If all of the deterministic behaviour of TorchIO is covered by our existing transforms I'd say this is fine, but if that's not fine then my first solution is maybe all we have and we'll have to trust the user to follow documentation stating how to use our wrappers.

Reply:
Maybe you guys can discuss how to get the traits/interfaces compliant between tio and monai and then tell me how you'd like me to implement/adjust it.

My personal opinion is the one I implemented, i.e. caching makes no sense with probabilistic application or - put differently - mark the transform as random if randomness is involved no matter where the randomness comes from.

In case you want to introduce the distinction between randomness inside the transformation and probabilistic transform application into MonAI's interfaces/traits I'm also happy to help with that.

Reply:
You might find this PR helpful for reference: [PR #5569](https://github.com/Project-MONAI/MONAI/pull/5569). There’s a discussion there that addresses application probability and inherent probability in more detail: [relevant discussion link](https://github.com/Project-MONAI/MONAI/pull/5569#discussion_r1030822630).